### DISCUSSION TITLE
Nexting a source on getRequestSources resulting observables

### DISCUSSION THREAD

**Discussion by pBouillon** [state-adapt/state-adapt#56] [Q&A] [answered]
Hi,

I would like to perform a side effect on the result of my call to a service method

I tried taking a look at [one of the examples](https://github.com/mfp22/stefanoslig-angular-14-ngrx-nx-realworld-example-app/blob/state-adapt/libs/articles/data-access/src/lib/%2Bstate/article-list/article-list.adapter.ts) but it relies on the deprecated `getHttpSources` method

Here is basically what I am trying to do:

```ts
@Injectable({ providedIn: 'root' })
export class MyService {
  readonly #todosService = inject(TodosService);

  // 👇 When this is nexted (1)
  readonly addAllTodos$ = new Source<Todo[]>('[Todos] Add all todos');
  readonly addAllTodosRequest = getRequestSources(
    '[Todos] addAllTodos$',
    // 👇 Performing the "side effect" (2)
    this.addAllTodos$.pipe(switchMap(() => this.#todosService.fetchAll()))
  );
  readonly addAllTodosSuccess$ = this.addAllTodosRequest.success$.pipe(
    map(({ payload }) => this.addAllTodos$.next(payload))
  );
  readonly addAllTodosError$ = this.addAllTodosRequest.error$.pipe(
    tap(({ payload }) => console.error('Got an error: ', payload))
  );

  readonly getAllTodos$ = new Source<void>('[Todos] Fetch all todos');

  readonly adapter = adapt(['todos-' + getId(), initialState, adapter], {
    // 👇 Hoping that this would trigger that (3)
    addAllTodos: this.addAllTodos$,
  });
}
```

However it does not seems to emit, what do I do wrong here?

**Comment by mfp22** [comment]
`adapt` returns a store, so the last property name should be `readonly store = adapt( // ...`

`adapt` will subscribe to the sources, so `addAllTodos$`, but that subscription doesn't go anywhere else. It stops there, because `addAllTodos$` doesn't pipe off of anything else. Just like in regular RxJS/Angular, an HTTP request will not be made until it is subscribed to. 

`fetchAll` seems like you're just fetching data from the server. When I saw `addAllTodos$`, I assumed it was saving new todos to the server. But it looks like you're just fetching them. So here is how I would do what you're doing:

```typescript
export class MyService {
  readonly #todosService = inject(TodosService);

  readonly addAllTodosRequest = getRequestSources(
    '[Todos] addAllTodos$',
    this.#todosService.fetchAll(),
  );

  readonly store = adapt(['todos-' + getId(), initialState, adapter], {
    addAllTodos: this.addAllTodosRequest.success$,
  });
}
```

**Reply by mfp22** [reply]
And don't forget to subscribe to some observable from the store as well. Either from a component with `toSignal` (Angular 16+), or `| async` in a template, or manually (not the best)

**Reply by pBouillon** [reply]
Thank you for the (detailed!) reply, indeed this is much better!

However, I would still like to perform that `console.error` as a side effect.
To tackle that issue, I took advantage of the `noop` property of the `sources` parameter of the `adapt` method (that I also saw in the example app):

```ts
readonly getAllTodo$ = new Source<void>('[Todo] Fetch all todo');

readonly #getAllTodoRequest = splitRequestSources(
  '[Todo] getAllTodo$',
  this.getAllTodo$.pipe(
    switchMap(() =>
      this.#todoService.fetchAll().pipe(toRequestSource('[Todo] getAllTodo$'))
    )
  )
);

readonly store = adapt(['todo-' + getId(), initialState, adapter], {
  addAllTodo: [this.#getAllTodoRequest.success$],
  noop: [
    this.#getAllTodoRequest.error$.pipe(tap(() => console.error('Loading Failed!'))),
  ],
});
```

But I wonder, `(1)` is this the "state-adapt-idiomatic" way of doing this?

Also, I noticed that the without the gymnastic using `splitRequestSources` and `toRequestSource`, this will not work once an error has been thrown and I had to put aside the usage of `getRequestSources`. `(2)` Is there a way for `getRequestSources` to continue working once an error has been thrown?

> PS: I did ask a lot of questions, if this is not the appropriate place to ask them, please let me know. I don't wan't to incorrectly use that space

**Reply by mfp22** [reply, marked as answer]
This is a great place for questions. I assume Google will index these at some point, and then more people can find answers than if I had a Discord server or something. 

Yes, this is a good StateAdapt way of doing things. `noop` is similar to `hold` in RxAngular, or an effect in NgRx/Effects. You don't need the array syntax, though. 

By the way, if you want different behavior with any of these source utilities, feel free to write your own. These aren't super important for StateAdapt. I just made these utilities because I wanted them often when using StateAdapt. If you look at the source code, you'll see that they really aren't doing very much. https://github.com/state-adapt/state-adapt/tree/main/libs/rxjs/src/lib/sources

**Reply by pBouillon** [reply]
Thanks a lot for the answers and your time!
I'll dig into the code to understand this a bit better :smile:
